core/utils/utils.go

```diff
@@ -581,3 +582,112 @@ func (da *dependentAwaiter) AddDependents(n int) {
 func (da *dependentAwaiter) DependentReady() {
 	da.wg.Done()
 }
+
+// FIFO queue that discards older items when it reaches its capacity.
+type BoundedQueue struct {
+	capacity uint
+	items    []interface{}
+	mu       *sync.RWMutex
+}
+
+func NewBoundedQueue(capacity uint) *BoundedQueue {
+	return &BoundedQueue{
+		capacity: capacity,
+		mu:       &sync.RWMutex{},
+	}
+}
+
+func (q *BoundedQueue) Add(x interface{}) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.items = append(q.items, x)
+	if uint(len(q.items)) > q.capacity {
+		excess := uint(len(q.items)) - q.capacity
+		q.items = q.items[excess:]
+	}
+}
+
+func (q *BoundedQueue) Take() interface{} {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if len(q.items) == 0 {
+		return nil
+	}
+	x := q.items[0]
+	q.items = q.items[1:]
+	return x
+}
+
+func (q *BoundedQueue) Empty() bool {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return len(q.items) == 0
+}
+
+func (q *BoundedQueue) Full() bool {
+	q.mu.RLock()
+	defer q.mu.RUnlock()
+	return uint(len(q.items)) >= q.capacity
+}
+
+type BoundedPriorityQueue struct {
+	queues     map[uint]*BoundedQueue
+	priorities []uint
+	capacities map[uint]uint
+	mu         *sync.RWMutex
+}
+
+func NewBoundedPriorityQueue(capacities map[uint]uint) *BoundedPriorityQueue {
```

Comment:
Can we get some tests for this?

Reply:
Sorry, appears the test got swallowed by a bad rebase.  Was able to recover it from my undo buffer.  Pushed.